jxl/src/container/mod.rs

```diff
@@ -0,0 +1,261 @@
+        bytes_left: Option<usize>,
+    },
+    InCodestream {
+        kind: BitstreamKind,
+        bytes_left: Option<usize>,
+    },
+    Done(BitstreamKind),
+}
+
+/// Structure of the decoded bitstream.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum BitstreamKind {
+    /// Decoder can't determine structure of the bitstream.
+    Unknown,
+    /// Bitstream is a direct JPEG XL codestream without box structure.
+    BareCodestream,
+    /// Bitstream is a JPEG XL container with box structure.
+    Container,
+    /// Bitstream is not a valid JPEG XL image.
+    Invalid,
+}
+
+impl ContainerParser {
+    const CODESTREAM_SIG: [u8; 2] = [0xff, 0x0a];
+    const CONTAINER_SIG: [u8; 12] = [0, 0, 0, 0xc, b'J', b'X', b'L', b' ', 0xd, 0xa, 0x87, 0xa];
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn kind(&self) -> BitstreamKind {
+        match self.state {
+            DetectState::WaitingSignature => BitstreamKind::Unknown,
+            DetectState::WaitingBoxHeader
+            | DetectState::WaitingJxlpIndex(..)
+            | DetectState::InAuxBox { .. } => BitstreamKind::Container,
+            DetectState::InCodestream { kind, .. } | DetectState::Done(kind) => kind,
+        }
+    }
+
+    pub fn feed_bytes(&mut self, input: &[u8]) -> Result<(), Error> {
+        let state = &mut self.state;
+        let buf = &mut self.buf;
+        buf.extend_from_slice(input);
+
+        loop {
+            match state {
+                DetectState::WaitingSignature => {
+                    if buf.starts_with(&Self::CODESTREAM_SIG) {
+                        // tracing::debug!("Codestream signature found");
+                        *state = DetectState::InCodestream {
+                            kind: BitstreamKind::BareCodestream,
+                            bytes_left: None,
+                        };
+                        continue;
+                    }
+                    if buf.starts_with(&Self::CONTAINER_SIG) {
+                        // tracing::debug!("Container signature found");
+                        *state = DetectState::WaitingBoxHeader;
+                        buf.drain(..Self::CONTAINER_SIG.len());
+                        continue;
+                    }
+                    if !Self::CODESTREAM_SIG.starts_with(buf)
+                        && !Self::CONTAINER_SIG.starts_with(buf)
+                    {
+                        // tracing::error!("Invalid signature");
+                        *state = DetectState::InCodestream {
+                            kind: BitstreamKind::Invalid,
+                            bytes_left: None,
+                        };
+                        continue;
+                    }
+                    return Ok(());
+                }
+                DetectState::WaitingBoxHeader => match ContainerBoxHeader::parse(buf)? {
+                    HeaderParseResult::Done { header, size } => {
+                        buf.drain(..size);
```

Comment:
I believe this is potentially quadratic if many small boxes are present and a large input is fed to the box decoder.
In general, unconditionally buffering the received input is best avoided too.

Reply:
Yeah, I just realized while thinking about how to reduce buffering. Addressed in c3e8a51.